File: BAEKJOON/그래프 탐색/벽 부수고 이동하기/벽 부수고 이동하기.cs
using System;
using System.Collections.Generic;

class Program
{
    struct State
    {
        public int X, Y, WallBroken;
        public State(int x, int y, int wallBroken)
        {
            X = x;
            Y = y;
            WallBroken = wallBroken;
        }
    }

    static int Bfs(int[][] maze, int n, int m)
    {
        // 이동 방향: 상, 하, 좌, 우
        int[][] directions = new int[][]
        {
            new int[] {-1, 0}, new int[] {1, 0},
            new int[] {0, -1}, new int[] {0, 1}
        };

        // BFS를 위한 큐 초기화: (x, y, 벽을 부쉈는지 여부)
        Queue<State> queue = new Queue<State>();
        queue.Enqueue(new State(0, 0, 0));
        
        // 방문 체크 리스트 초기화
        bool[,,] visited = new bool[n, m, 2];
        visited[0, 0, 0] = true;

        // 거리 리스트 초기화
        int[,,] distance = new int[n, m, 2];
        distance[0, 0, 0] = 1;

        while (queue.Count > 0)
        {
            State current = queue.Dequeue();
            int x = current.X;
            int y = current.Y;
            int wallBroken = current.WallBroken;

            // 도착 지점에 도달했을 때
            if (x == n - 1 && y == m - 1)
            {
                return distance[x, y, wallBroken];
            }

            // 4방향 탐색
            foreach (var direction in directions)
            {
                int nx = x + direction[0];
                int ny = y + direction[1];

                if (nx >= 0 && nx < n && ny >= 0 && ny < m)
                {
                    if (maze[nx][ny] == 0 && !visited[nx, ny, wallBroken])
                    {
                        // 벽이 아니고, 방문하지 않은 위치일 때
                        visited[nx, ny, wallBroken] = true;
                        distance[nx, ny, wallBroken] = distance[x, y, wallBroken] + 1;
                        queue.Enqueue(new State(nx, ny, wallBroken));
                    }

                    if (maze[nx][ny] == 1 && wallBroken == 0 && !visited[nx, ny, 1])
                    {
                        // 벽이지만, 아직 벽을 부수지 않은 경우
                        visited[nx, ny, 1] = true;
                        distance[nx, ny, 1] = distance[x, y, wallBroken] + 1;
                        queue.Enqueue(new State(nx, ny, 1));
                    }
                }
            }
        }

        return -1; // 도착 지점에 도달할 수 없는 경우
    }

    static void Main(string[] args)
    {
        string[] input = Console.ReadLine().Split();
        int n = int.Parse(input[0]);
        int m = int.Parse(input[1]);

        int[][] maze = new int[n][];
        for (int i = 0; i < n; i++)
        {
            string row = Console.ReadLine();
            maze[i] = new int[m];
            for (int j = 0; j < m; j++)
            {
                maze[i][j] = row[j] - '0';
            }
        }

        Console.WriteLine(Bfs(maze, n, m));
    }
}

File: BAEKJOON/그래프 탐색/벽 부수고 이동하기/벽 부수고 이동하기.py
from collections import deque

def bfs(maze, n, m):
    # 이동 방향: 상, 하, 좌, 우
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    # BFS를 위한 큐 초기화: (x, y, 벽을 부쉈는지 여부)
    queue = deque([(0, 0, 0)])  # 시작점 (0, 0)에서 시작, 벽을 부수지 않은 상태 (0)
    
    # 방문 체크 리스트 초기화
    visited = [[[False] * 2 for _ in range(m)] for _ in range(n)]
    visited[0][0][0] = True  # 시작점 방문 표시
    
    # 거리 리스트 초기화
    distance = [[[0] * 2 for _ in range(m)] for _ in range(n)]
    distance[0][0][0] = 1  # 시작점의 거리
    
    while queue:
        x, y, wall_broken = queue.popleft()  # 현재 위치와 벽을 부쉈는지 여부를 큐에서 꺼냄
        
        # 도착 지점에 도달했을 때
        if x == n - 1 and y == m - 1:
            return distance[x][y][wall_broken]  # 현재까지의 거리 반환
        
        # 4방향 탐색
        for dx, dy in directions:
            nx, ny = x + dx, y + dy  # 다음 위치 계산
            
            if 0 <= nx < n and 0 <= ny < m:  # 미로 범위 내에 있을 때
                if maze[nx][ny] == 0 and not visited[nx][ny][wall_broken]:
                    # 벽이 아니고, 방문하지 않은 위치일 때
                    visited[nx][ny][wall_broken] = True  # 방문 표시
                    distance[nx][ny][wall_broken] = distance[x][y][wall_broken] + 1  # 거리 업데이트
                    queue.append((nx, ny, wall_broken))  # 큐에 추가
                
                if maze[nx][ny] == 1 and wall_broken == 0 and not visited[nx][ny][1]:
                    # 벽이지만, 아직 벽을 부수지 않은 경우
                    visited[nx][ny][1] = True  # 방문 표시
                    distance[nx][ny][1] = distance[x][y][wall_broken] + 1  # 거리 업데이트
                    queue.append((nx, ny, 1))  # 큐에 추가 (벽을 부순 상태로)
    
    return -1  # 도착 지점에 도달할 수 없는 경우

# 입력 받기
n, m = map(int, input().split())  # N x M 크기의 미로
maze = [list(map(int, input().strip())) for _ in range(n)]  # 미로의 상태

# 결과 출력
print(bfs(maze, n, m))  # BFS를 수행하여 결과 출력
